### PR TITLE
[tests][dask] fix workers without data test (fixes #5537)

### LINKED ISSUE #5537: [ci] [dask] fix test on "only one worker has data" training
## Description

In #5510, we observed that the Dask tests were getting stuck and timing out after 60 minutes.

Narrowed that down to a single unit test, which is currently skipped.

https://github.com/microsoft/LightGBM/blob/0c0eb2a6a1ba111f0896dd16a579342fa16f7629/tests/python_package_test/test_dask.py#L1506-L1507

The testing work from #5510 should be continued.... that test should be re-enabled and its underlying cause should be identified and fixed.

## Reproducible example

Uncomment the test linked above and push a commit to a pull request in LightGBM.

## Environment info

Latest version of LightGBM (https://github.com/microsoft/LightGBM/commit/0c0eb2a6a1ba111f0896dd16a579342fa16f7629).

## Additional Comments

Especially see this comment describing a possible root cause of timeout issues: https://github.com/microsoft/LightGBM/pull/5510#issuecomment-1263525112.

### PR DESCRIPTION
**Pull request by jmoralez**
Changes `tests/python_package_test/test_dask.py::test_training_succeeds_even_if_some_workers_do_not_have_any_data` to have three workers and data in only two of them.

Fixes #5537.

### CODE REVIEW

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for working on this!

Before I review, could you please explain, either in the description or in a comment here, what the root cause of the timeouts observed in #5537 were and *how* this fixes it?

The diff has a lot more changes in it than I'd expected from the description "have three workers and data in only two of them". For example, it's not obvious to me how that type of change would require bringing in new imports from `sklearn.metrics`.

I'm especially interested in the change where we're now passing an explicit list of `workers=` to `client.persist()`. Was the root problem just that "`collection_to_single_partition()` was resulting in, for example, `dX` being all on the first worker and `dy` being all on the second worker"? If so, couldn't the same effect be accomplished with the existing 2-worker cluster?

**Comment by jmoralez** [comment]
> what the root cause of the timeouts observed in https://github.com/microsoft/LightGBM/issues/5537 were

I couldn't reproduce the timeouts locally.

> how this fixes it

The purpose of the test was to check that in a distributed training setting, having a worker without data wouldn't cause a problem, that's why I changed it to having three workers and only two with data, so distributed training is performed (because of the two workers having data), whereas previously it was probably falling back to just local training, and the definition of "succeeds" is that we achieve a good metric value (hence the `sklearn.metrics` import).

**Review by jameslamb — APPROVED** [review]
> previously it was probably falling back to just local training

OHHHHHHHHHHHH. Thank you so much for that explanation. Makes sense to me!

Let's merge this then 😁

**Comment by jameslamb** [comment]
Will merge this whenever we get CI working again 😭 

Hopefully if/when #5514 is merged.

**Comment by jameslamb** [comment]
thanks so much for fixing this @jmoralez !
